[incubator/vault] Initial commit of vault chart

jpds opened the pull request.

jpds committed:
[incubator/vault] Initial commit of vault chart.

mumoshu commented on incubator/vault/values.yaml:
```diff
@@ -0,0 +1,44 @@
+    #storage:
+    #  consul:
+    #    address: ""
+    #    path: ""
```
@jpds Hi, thanks for the great PR!
It could sound a nit but can we make this part uncommented and modified like:

```yaml
# A YAML representation of a final vault config.json file.
# See https://www.vaultproject.io/docs/configuration/index.html for more information
vault:
  listener:
    tcp:
      address: '0.0.0.0:8200'
      tls_disable: 1
  storage:
    consul:
      address: ""
      path: ""
```

so that for a basic use-case we can just override `consul.address` via the `--set` flag like `helm install stable/vault --set vault.storage.consul.address="myconsul-svc-name"`?

jpds committed:
[incubator/vault] Use consul backend by default.

jpds committed:
[incubator/vault] README.md: Added S3 example.

jpds committed:
[incubator/vault] Made deployment root read-only.

jpds committed:
[incubator/vault] configmap: Updated vault config name.

jpds committed:
[incubator/vault] README: Added consul port.

jpds committed:
[incubator/vault] Included links to vault config docs.

jpds commented:
Hello @mumoshu, thanks for the review - I've gone ahead and added those suggestions in and an alternative config in README too.

jpds committed:
[incubator/vault] Put /root/ under emptyDir so that the token can be saved on
read-only rootfs.

jpds committed:
[incubator/vault] README: Added "Using Vault" section.

jpds committed:
[incubator/vault]: Uncomment example consul section to avoid double 'storage'
section conflict.

jpds committed:
[incubator/vault] README: Added consul.path to example.

kachkaev commented:
This is awesome!

@jpds what do you think of adding [vault-ui](https://github.com/djenriquez/vault-ui) as an optional component that would be available through ingress? Setting `ui.enabled` will turn it on and then adding things like `ui.ingress.xxx` will tweak the details. Or should there be a separate chart called `vault-ui` that depends on this chart (like wordpress depends on mariadb)?

jpds commented:
@kachkaev Hello, I'd rather keep them separate (versioning and there seem to be a small handful of vault-ui projects out there).

I've pushed a PR for vault-ui to https://github.com/kubernetes/charts/pull/1406

unguiculus reviewed (changes requested):
Thanks for the PR. You didn't get standard labels quite right. In your other PR(#1406), they are fine.

jpds committed:
[incubator/vault] Corrected labels for manifests.

jpds commented:
@unguiculus Ah, this is a `helm create` 2.4 vs 2.5 thing, I've fixed it now. Thanks.

unguiculus commented:
/ok-to-test

jpds commented:
@unguiculus Do you know if there's a way to specify a values.yaml just for testing purposes?

The test is failing as there's no default storage backend enabled and I can't enable that as then vault will complain that there are two storage backends when people specify their own.

jpds committed:
[incubator/vault] Chart.yaml: Fixed logo -> icon.

unguiculus commented:
A solution for test values is currently being worked on. However, in general, it should be possible to install a chart with default values. Do you think that would be somehow feasible?

jpds committed:
[incubator/vault] Github username as name.

jpds commented:
If a backend isn't provided or if two backends are provided - vault just errors out. Hence why I can't provide a default simple file backend in the YAML, while letting people use their own backend.

jpds committed:
[incubator/vault] README: We don't specify a backend.

mumoshu commented:
@jpds Not elegant at all but how about introducing the new key `config.json` to accept json like:

```yaml
config:
  json:|
    {
      "listener": {
        "tcp": {
          "address": "0.0.0.0:8200",
          "tls_disable": 1,
      },
      "storage": ...
    }
```

so that we can override the whole config.json?

Also, I guess you can add a conditional dependency to the consul chart via requirements.yaml while making the default `config.json` to assume settings correspond to the consul chart.

lachie83 commented on incubator/vault/values.yaml:
```diff
@@ -0,0 +1,44 @@
+    tcp:
+      address: '0.0.0.0:8200'
+      tls_disable: 1
+  storage:
```
Can you call out a the URL of where to find a list of supported backends in the config?

ipedrazas commented:
What I have is a flag that enables the dev mode, which doesn't need backend.

```
          {{- if .Values.vault.mode.dev }}
          command: ["vault", "server", "-dev"]
          {{- else }}
          command: ["vault", "server", "-config", "/conf/config.json"]
          {{- end }}
```

This allows to install the chart and the pod goes into running state just fine.

unguiculus commented:
Marking as stale. Please update within the next week.

jacobkiers commented:
@jpds Are you still willing to finish this?

If not, I am interested in seeing this through. I'm new to creating Helm charts though, so I might need a little guidance from a reviewer.

unguiculus commented:
@jacobkiers Thanks for jumping in. We are happy to assist. You may also ping us on Slack.

jacobkiers commented:
First  a small question: how should I go about this? Obviously, I cannot push to the branch of @jpds. Should I create a new PR or something?

Commit:
[incubator/vault] Allow dev mode

As suggested by @ipedrazas.

Signed-off-by: Jacob Kiers <kiers@comandi.nl>

Commit:
[incubator/vault] Add some more storage examples

Signed-off-by: Jacob Kiers <kiers@comandi.nl>

unguiculus commented:
@jacobkiers You could actually push to his branch, but it is probably better to fork the PR and create a new one.

@jpds: Last call. Are you willing to continue with this? Otherwise I'll close this PR so @jacobkiers can pick it up.

jpds commented:
Hello,

I do plan to work on this, I've been very busy recently but should have time to work on this next week.

Thanks,

-------- Message d'origine --------
On 4 août 2017 à 15:33, Reinhard Nägele a écrit :

> [@jacobkiers](https://github.com/jacobkiers) You could actually push to his branch, but it is probably better to fork the PR and create a new one.
>
> [@jpds](https://github.com/jpds): Last call. Are you willing to continue with this? Otherwise I'll close this PR so [@jacobkiers](https://github.com/jacobkiers) can pick it up.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, [view it on GitHub](https://github.com/kubernetes/charts/pull/1397#issuecomment-320249922), or [mute the thread](https://github.com/notifications/unsubscribe-auth/AbzuO9qWGFI6lF-ZhbP5M8oSpRioqrMRks5sUx2XgaJpZM4OInt1).

jacobkiers commented:
@jpds That's great! I have some small changes already. I'll push my branch
to my own repo so that you don't need to do work that's already done.

Edit: removed mail cruft.

Commit:
[incubator/vault] Use revision for vault config map

Using the revision in the config map allows upgrading the vault
configuration.

Without this, the deployment of vault is never recreated after updating
the configuration.

There is a disadvantage though: after upgrading, vault needs to be
unsealed again, because the pod is restarted

Signed-off-by: Jacob Kiers <kiers@comandi.nl>

Commit:
[incubator/vault] Make the notes correct

Signed-off-by: Jacob Kiers <kiers@comandi.nl>

jacobkiers commented:
@jpds Please see my changes at https://github.com/comandi/charts/commits/vault-initial-chart.

It has two notable changes:
* I implemented the suggestion by @ipedrazas to allow for a dev-server deployment. When that is enabled, all other configuration is ignored.
* The config map containing the Vault config is now tied to the Helm revision. That make sure that when the config is upgraded, de pod will be redeployed with the new configuration. It has the disadvantage that in that case Vault needs to be unsealed again. If there is a better way to do so, feel free to implement that.

I noticed one other thing that I didn't get around to yet: when the liveness and readyness probes are enabled, the pod went into a `CrashLoopBackoff` state. But I don't yet have enough knowledge to find out why.

jacobkiers commented:
@unguiculus If we want to incorporate some of these changes, then I should probably sign the CLA as well. I'm not sure though whether to do so as an employee or as an individual. I do know that my employer has not yet signed the CLA though.

unguiculus commented:
@jacobkiers I guess you'll have to find out by yourself how you want to do it.

jacobkiers commented:
My employer has signed the CLA, so my contributions should be fine to use.

tuannvm commented:
@jacobkiers @jpds @unguiculus I would love to contribute as well.

@jacobkiers 
> The config map containing the Vault config is now tied to the Helm revision. That make sure that when the config is upgraded, de pod will be redeployed with the new configuration. It has the disadvantage that in that case Vault needs to be unsealed again. If there is a better way to do so, feel free to implement that.

the helm documentation suggests us to use `checksum` to enable roll deployments [here](https://github.com/kubernetes/helm/blob/master/docs/charts_tips_and_tricks.md#automatically-roll-deployments-when-configmaps-or-secrets-change). Furthermore, Helm will not create a lot of configmap with the equivalent revision numbers, that could save us from cleaning up job.

Feel free to take a look on my suggestion [here](https://github.com/tuannvm/charts/commits/tuan-vault-implementation)

jpds committed:
[incubator/vault] Updated image.tag to 0.8.1.

jpds committed:
[incubator/vault] Added vaultmode.dev to parameter table.

jpds commented:
This should be good to go now.

erickt commented on incubator/vault/Chart.yaml:
```diff
@@ -0,0 +1,13 @@
+description: A Helm chart for Vault, a tool for managing secrets
+name: vault
+version: 0.1.0
+appVersion: 0.7.3
```
This should be 0.8.1, to match the tagged version.

erickt commented on incubator/vault/values.yaml:
```diff
@@ -0,0 +1,61 @@
+# Only used to enable dev mode. When in dev mode, the rest of this configuration
+# is ignored. See https://www.vaultproject.io/intro/getting-started/dev-server.html
+# for more information.
+  dev: true
```
Could you document in the README that this setup is insecure by default, and what is the configuration needed to setup TLS on an internal cluster that cannot use Lets Encrypt?

ipedrazas replied:
Running Vault in dev mode is insecure by design.

erickt replied:
Understood, but given the fact that vault is supposed to store secrets, I think it's worth the effort to keep people from running an insecure vault.

Now that I think about it, we should also add a warning in the status after we install vault if we are in dev mode.

erickt commented on incubator/vault/values.yaml:
```diff
@@ -0,0 +1,61 @@
+# A YAML representation of a final vault config.json file.
+# See https://www.vaultproject.io/docs/configuration/ for more information.
+vaultmode:
+# Only used to enable dev mode. When in dev mode, the rest of this configuration
```
What do you mean by "rest of this configuration is ignored"? Are you referring to all the options under `vaultmode`? Also, I think it might read a little better if you had:

```yaml
vault:
    dev: true
    config:
        listener:
            ...
        storage:
            ...
```

That gives a more of a namespace to add other vault options that are outside the config file.

erickt commented on incubator/vault/README.md:
```diff
@@ -0,0 +1,69 @@
+
+> See https://www.vaultproject.io/docs/configuration/ for more information.
+
+## Installing the Chart
```
Hashicorp recommends running Vault on a [dedicated instance](https://www.vaultproject.io/guides/production.html) to reduce the risk of the machine running Vault getting rooted, and thus allowing an attacker to get to unencrypted secrets. Can you add links to documentation like [kops's instance groups](https://github.com/kubernetes/kops/blob/master/docs/instance_groups.md) and an example config on how to isolate pods to those nodes?

jpds replied:
I think running an entire VM instance just for the one container that is running (a rootfs read-only) vault is a little overkill at that point. :-)

erickt replied:
The risk here is that if Vault is running on the same machine as some other container that's able to get root on the system, it can read the Vault secrets directly out of memory. I've seen some Vault setups that go even more extreme than machine-isolation. For example, Nike created [Cerberus](http://engineering.nike.com/cerberus) that wraps around Vault, and they recommend isolating their setup off in a [dedicated AWS account](http://engineering.nike.com/cerberus/docs/administration-guide/creating-an-environment). I don't think we should hold up this chart for this documentation, but I think it'd be worthwhile adding it in the future.

jpds committed:
[incubator/vault] Placed everything under vault section for config.

jpds committed:
[incubator/vault] Chart.yaml: Updated version.

jpds committed:
[incubator/vault] values.yaml: Moved comment to the right place.

jpds committed:
[incubator/vault] README.md: Final tweaks for merge review.

jacobkiers commented on incubator/vault/values.yaml:
```diff
@@ -0,0 +1,62 @@
+  # section below is not used to configure Vault. See
+  # https://www.vaultproject.io/intro/getting-started/dev-server.html for more
+  # information.
+  dev: true
```
Why is dev-mode enabled by default?

jpds replied:
So the CI tests can pass with the pod becoming `Ready`.

jacobkiers replied:
Sounds fair.

jacobkiers commented on incubator/vault/templates/deployment.yaml:
```diff
@@ -0,0 +1,52 @@
+        {{- end }}
+        ports:
+        - containerPort: {{ .Values.service.port }}
+        livenessProbe:
```
I had trouble getting vault up-and-running with the liveness (and/or readyness probe) enabled, because the time between starting vault and unsealing it was too long.

@jpds Did you test this also with a non-dev config? If so, how did you solve that?

jpds replied:
Yes, I've used this with before consul and etcd as the backends. Just spun up a new etcd cluster to test it. Didn't have an issue with it starting or unsealing.

jpds replied:
Stock `incubator/etcd` chart + this as `vault-values.yaml`:

```
vault:
  dev: false
  config:
    storage:
      etcd:
        address: "http://my-release-etcd-0.my-release-etcd.default:2379,my-release-etcd-1.my-release-etcd.default:2379,my-release-etcd-2.my-release-etcd.default:2379"
```

jacobkiers replied:
Thanks for verifying!

Then it seems there was an error at my end. I'm not that familiar with these probes yet...

jpds replied:
I would take a look at the logs from the pod and see if there's a config error.

jacobkiers commented:
Seems good to go to me as well.

erickt commented on incubator/vault/values.yaml:
```diff
@@ -0,0 +1,62 @@
+  enabled: false
+  # Used to create Ingress record (should used with service.type: ClusterIP).
+  hosts:
+    - chart-example.local
```
This should probably be commented out.

erickt reviewed (commented):
One minor nit, but beyond that, it looks good to go for me!

I'd love more documentation and warnings about running in dev mode, but we can always add that later.

erickt commented:
@jpds: Thought of something else while we wait for final review. Could you add anti-affinity to the Deployment, like [this](https://github.com/kubernetes/charts/blob/master/stable/concourse/templates/worker-statefulset.yaml#L79-L88)? That will tell kubernetes to not schedule the Vault replica pods on the same machine.

jpds committed:
[incubator/vault] values.yaml: Comment out initial ingress definition.

jpds committed:
[incubator/vault] deployment: Added podAntiAffinity definition.

jpds commented:
@erickt This anti-affinity definition seems to (strangely) cause a `CrashLoopBackOff`?

jpds commented:
@unguiculus Apart from the last commit failing - this one should be good to go and it's only your review that's pending.

jpds committed:
[incubator/vault] Updated to 0.8.2.

jpds committed:
[incubator/vault] deployment: Have server listen on [::] for health check (in both modes).

jpds committed:
[incubator/vault] Added 'vault status' test case.

unguiculus commented on incubator/vault/README.md:
```diff
@@ -0,0 +1,70 @@
+
+## Prerequisites Details
+
+* Kubernetes 1.5
```
We support the current and the previous release, thus: 1.6+

unguiculus commented on incubator/vault/templates/_helpers.tpl:
```diff
@@ -0,0 +1,16 @@
+{{/*
+Expand the name of the chart.
+*/}}
+{{- define "name" -}}
```
Use namespaced templates: `vault.name` and `vault.fullname`

See #1785 and https://github.com/kubernetes/helm/pull/2845.

unguiculus commented on incubator/vault/values.yaml:
```diff
@@ -0,0 +1,62 @@
+      #  access_key: ""
+      #  secret_key: ""
+      #  endpoint: "" # When not using AWS S3
+resources:
```
Don't specify defaults for resources.

kachkaev replied:
Hi @unguiculus. Why so? We have a discussion on default `memoryLimit` for traefik in https://github.com/kubernetes/charts/issues/2098, knowing your opinion would be useful.

unguiculus replied:
In general, we recommend to leave this a conscious choice for the user. Users should be able to test a chart out of the box on a system with little resources such as Minikube. If you think you have to specify them (and they are really low anyways in this case), then leave it as it is.

jpds committed:
[incubator/vault] Updated README per review.

jpds committed:
[incubator/vault] _helpers.tpl: Use namespaced template names.

unguiculus commented:
/lgtm
